src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    partial class JsonDocument
+    {
+        private struct CustomDb : IDisposable
+        {
+            private const int SizeOrLengthOffset = 4;
+            private const int NumberOfRowsOffset = 8;
+
+            internal int Length;
+            private byte[] _rentedBuffer;
+            private ArrayPool<byte> _pool;
```

Comment:
Do we need to work on arbitrary pools, rather than just using `ArrayPool<T>.Shared`?  Not having to store the pool will make these structs smaller, plus there's a much higher likelihood that Rent/Return can get devirtualized and inlined with `ArrayPool<T>.Shared.Rent(…)/Return(…)`.